File: languages/python/src/totallman/converter.py
from __future__ import annotations

import unicodedata

from totallman._embedded_lists import get_list, get_max_word_count
from totallman._unicode import case_fold, is_letter_or_mark


def to_tallman(text: str | None, list_id: str = "DEFAULT") -> str:
    """Convert medication names in *text* to Tall Man lettering.

    Implements the canonical character-by-character greedy longest-match
    algorithm (spec §7). Raises TallmanError for an unknown list_id.

    Args:
        text: The input string. None or empty string returns "".
        list_id: Tall Man list to use. Defaults to "DEFAULT".

    Returns:
        The input string with recognised drug names converted to Tall Man form.
    """
    if not text:
        return ""

    # Step 1 — NFC-normalise the whole input.
    normalised = unicodedata.normalize("NFC", text)

    # Step 2 — look up the list (raises TallmanError for unknown list_id).
    dictionary = get_list(list_id)
    max_words = get_max_word_count(list_id)

    result: list[str] = []
    chars = normalised
    n = len(chars)
    i = 0

    while i < n:
        if is_letter_or_mark(chars[i]):
            # Extract the word (contiguous letters/marks).
            word, i = _extract_word(chars, i, n)

            # Try single-word match first.
            best_match: str | None = dictionary.get(case_fold(word))
            best_end = i

            # Greedy lookahead for multi-word / hyphenated entries.
            look = i
            words_in_pattern = 1
            pattern = word

            while words_in_pattern < max_words and look < n and chars[look] in (" ", "-"):
                sep = chars[look]
                look += 1
                if look < n and is_letter_or_mark(chars[look]):
                    next_word, look = _extract_word(chars, look, n)
                    pattern = pattern + sep + next_word
                    words_in_pattern += 1
                    multi_match = dictionary.get(case_fold(pattern))
                    if multi_match is not None:
                        best_match = multi_match
                        best_end = look
                else:
                    break

            if best_match is not None:
                result.append(best_match)
                i = best_end
            else:
                result.append(word)
                # i already advanced by _extract_word
        else:
            result.append(chars[i])
            i += 1

    return "".join(result)


def _extract_word(text: str, start: int, n: int) -> tuple[str, int]:
    """Return (word, end_index) for the letter/mark run beginning at start."""
    i = start
    while i < n and is_letter_or_mark(text[i]):
        i += 1
    return text[start:i], i

File: languages/js/src/converter.ts
import { EMBEDDED_LISTS } from './embedded-lists.js';
import { nfc, caseFold, isLetterOrMark, charCount } from './unicode.js';
import { TallmanError } from './tallman-error.js';

function getList(listId: string) {
  const list = Object.hasOwn(EMBEDDED_LISTS, listId) ? EMBEDDED_LISTS[listId] : undefined;
  if (list === undefined) {
    const available = Object.keys(EMBEDDED_LISTS).sort().join(', ');
    throw new TallmanError(`Unknown Tallman list ID: '${listId}'. Available lists: ${available}.`);
  }
  return list;
}

export function toTallman(text: string, listId: string = 'DEFAULT'): string {
  if (!text) return '';

  const normalized = nfc(text);
  const { entries, maxWords } = getList(listId);

  let result = '';
  let i = 0;
  const n = normalized.length;

  while (i < n) {
    if (isLetterOrMark(normalized, i)) {
      // Extract the contiguous letter/mark word.
      const wordStart = i;
      while (i < n && isLetterOrMark(normalized, i)) {
        i += charCount(normalized, i);
      }
      const word = normalized.slice(wordStart, i);

      // Single-word match attempt.
      const wordKey = caseFold(word);
      let bestMatch: string | undefined = Object.hasOwn(entries, wordKey)
        ? entries[wordKey]
        : undefined;
      let bestEnd = i;

      // Greedy multi-word lookahead — bounded by per-list maxWords.
      // Separators: space ' ' and hyphen '-' only (spec section 3.4).
      let look = i;
      let wordsInPattern = 1;
      let pattern = word;

      while (wordsInPattern < maxWords && look < n) {
        const sep = normalized[look];
        if (sep !== ' ' && sep !== '-') break;
        pattern += sep;
        look++;

        if (look < n && isLetterOrMark(normalized, look)) {
          const nextWordStart = look;
          while (look < n && isLetterOrMark(normalized, look)) {
            look += charCount(normalized, look);
          }
          pattern += normalized.slice(nextWordStart, look);
          wordsInPattern++;

          const multiKey = caseFold(pattern);
          const multiMatch = Object.hasOwn(entries, multiKey) ? entries[multiKey] : undefined;
          if (multiMatch !== undefined) {
            bestMatch = multiMatch;
            bestEnd = look;
          }
        } else {
          break;
        }
      }

      result += bestMatch ?? word;
      i = bestEnd;
    } else {
      // Non-letter: copy verbatim, handling surrogate pairs correctly.
      const cp = normalized.codePointAt(i) ?? 0;
      result += String.fromCodePoint(cp);
      i += cp > 0xffff ? 2 : 1;
    }
  }

  return result;
}

export function availableLists(): ReadonlyArray<string> {
  return Object.keys(EMBEDDED_LISTS).sort();
}

export function listVersion(listId: string): string {
  return getList(listId).version;
}
